src/requestManager.cpp

```diff
@@ -164,13 +164,43 @@ void CRequestManager::AskFor(const CInv &obj, CNode *from, unsigned int priority
 // Get these objects from somewhere, asynchronously.
 void CRequestManager::AskFor(const std::vector<CInv> &objArray, CNode *from, unsigned int priority)
 {
-    unsigned int sz = objArray.size();
-    for (unsigned int nInv = 0; nInv < sz; nInv++)
+    // In order to maintain locking order, we must lock cs_objDownloader first and before possibly taking cs_vNodes.
+    // Also, locking here prevents anyone from asking again for any of these objects again before we've notified the
+    // request manager of them all. In addition this helps keep blocks batached and requests for batches of blocks
+    // in a better order.
+    LOCK(cs_objDownloader);
+    for (auto &inv : objArray)
     {
-        AskFor(objArray[nInv], from, priority);
+        AskFor(inv, from, priority);
     }
 }
 
+void CRequestManager::AskForDuringIBD(const std::vector<CInv> &objArray, CNode *from, unsigned int priority)
+{
+    // add from this node first so that they get requested first.
+    LOCK(cs_objDownloader);
+    AskFor(objArray, from, priority);
+
+    // Must lock cs_objDownloader here and before cs_vNodes to maintain proper locking order.
+    LOCK(cs_vNodes);
+    for (CNode *pnode : vNodes)
+    {
+        if (pnode == from)
+            continue;
+
+        AskFor(objArray, pnode, priority);
```

Comment:
it looks here like you are asking for every block from every connected node?

Reply:
yes, we have to do this here.  We need to add a block source for every peer otherwise we can't do re-requests.  But what I should do is (1) only add sources for NETWORK nodes and then (2) perhaps filter for peers that have block availablility , if I can actually do that here, and if I can get the correct block availability ... i'll work on this along with getting the correct block availability for each peer rather than assuming they all have the blocks.

Reply:
was wondering, if a node as NODE_NETWORK set and its current height (CH) is higher than our current best chain tip (CT), what is the reason why it wouldn't be able to provide to us any blocks with height H belonging to the (CT, CH]?